### PR TITLE
Add a CPU lowering pass

### PR DESCRIPTION
**Pull request by shintaro-iwasaki**
Per discussion in #24, this PR creates a reference-CPU backend for `triton-shared` using the standard MLIR->LLVM lowering passes.

- Without any changes in the Triton runtime, we can run `triton-shared` on CPU from Python-Triton code.
  - Potentially it is useful for testing.
- A very basic kernel (see `reduce.py`) runs and succeeds.
- Many kernels do not run because of a lack of LLVM lowering and other problems.
  - For example, `memref::tensorStore`->LLVM seems not implemented, so many kernels are not supported at this point.

Help from the community and triton-shared core developers is needed to improve and maintain it.  I'd appreciate your feedback and suggestions.

Note: compared with the PR I used in the discussion, this PR version can run a compute kernel (`reduce.py`, as an example) by using MLIR's `CRunnerUtils.h`.

(EDIT: let me understand if I can sign this CLA)

### CODE REVIEW

**Comment by nhat-nguyen** [comment]
Thank you so much @shintaro-iwasaki! I'll start taking a look 😄

File: python/__init__.py
Comment (nhat-nguyen) [inline python/__init__.py:17]: this should be `raise`

File: python/__init__.py
Comment (nhat-nguyen) [inline python/__init__.py:25]: nit: i think this can be rewritten using `os.path.join` similarly to others to make things consistent

File: python/__init__.py
Comment (nhat-nguyen) [inline python/__init__.py:37]: nit: `subprocess.check_call` always returns 0 or raises an exception otherwise, we can remove all of the asserts😄

File: python/__init__.py
Comment (nhat-nguyen) [inline python/__init__.py:97]: nit: i think we can use `Path(src_path).write_text(llir)` like others for consistency

File: python/__init__.py
Comment (nhat-nguyen) [inline python/__init__.py:349]: nit: i think we can return `cache_path` early if it is `not None` to reduce the nesting level below (we have a big string below so will help with readability a bit)

File: python/__init__.py
Comment (nhat-nguyen) [inline python/__init__.py:389]: nit: similar to the assert comment above, this is not needed

File: python/__init__.py
Comment (nhat-nguyen) [inline python/__init__.py:400]: this is just for my own curiosity, what would happen if we put `py_src` directly and set `binary=False` instead of reading `dst_path` again?

File: python/ExecutionEngine/version.txt
Comment (nhat-nguyen) [inline python/ExecutionEngine/version.txt:1]: thanks for adding the commit hash for these files 😄

File: .github/workflows/integration-tests.yml
Comment (nhat-nguyen) [inline .github/workflows/integration-tests.yml:19]: once we merge this, i think it would be a good idea to also run this on triton main so that we know of any regressions

**Review by nhat-nguyen — COMMENTED** [review]
@shintaro-iwasaki I left a few nit comments, other than that, everything looks good! Thanks again for contributing this. @manbearian mentioned that he couldn't get tensor store to work, so I'll be investigating that separately once we merge this.

**Comment by nhat-nguyen** [comment]
@kshama-msft Would you mind helping Shintaro with the CLA question? Thank you!

**Comment by kshama-msft** [comment]
@shintaro-iwasaki if the CLA has been generated by the github bot then feel free to sign it. You should review with your internal open-source contribution legal teams to confirm that it is okay for you to sign.

**Comment by shintaro-iwasaki** [comment]
Thanks a lot for your review, @nhat-nguyen!
I am still figuring out the CLA thing with the legal team. Please give me some more time. Once it's cleared, let me fix all the issues pointed out.

**Comment by shintaro-iwasaki** [comment]
@microsoft-github-policy-service agree [company="Meta Platforms, Inc."]

**Comment by shintaro-iwasaki** [comment]
@microsoft-github-policy-service agree company="Meta Platforms, Inc."

**Comment by shintaro-iwasaki** [comment]
The legal thing has been cleared on my side. Let me fix the issues pointed out and update the PR by the end of today.

File: python/__init__.py
Comment (shintaro-iwasaki) [reply python/__init__.py]: Thanks! Yes it should be `raise`. Updated in the latest commit.

File: python/__init__.py
Comment (shintaro-iwasaki) [reply python/__init__.py]: Changed it to `os.path.join()`.

File: python/__init__.py
Comment (shintaro-iwasaki) [reply python/__init__.py]: Thanks! I didn't know that. Removed all of the `ret == 0` patterns for `subprocess.check_call()`.

File: python/__init__.py
Comment (shintaro-iwasaki) [reply python/__init__.py]: Agree. Updated it to `write_text()`. I believe all the others use this in `__init__.py`.

File: python/__init__.py
Comment (shintaro-iwasaki) [reply python/__init__.py]: Changed the code structure to return early when `cache_path is not None`.

File: python/__init__.py
Comment (shintaro-iwasaki) [reply python/__init__.py]: Updated!

File: python/__init__.py
Comment (shintaro-iwasaki) [reply python/__init__.py]: Thank you. This is a very good point. It works by just passing `py_src` with `binary=False`. Updated the code.

**Comment by shintaro-iwasaki** [comment]
Thanks! I updated the commit as well as rebased onto the latest `main`. Please let me know any further suggestions!

(Do you want the PR author (me) to resolve the comments? Or should it be done by reviewers?)

**Comment by nhat-nguyen** [comment]
> Thanks! I updated the commit as well as rebased onto the latest `main`. Please let me know any further suggestions!
> 
> (Do you want the PR author (me) to resolve the comments? Or should it be done by reviewers?)

Thank you @shintaro-iwasaki, everything looks good! We usually let the PR authors resolve the comments but let me know if you'd prefer otherwise. Once the pipeline runs finish, I can merge in the changes. 😄

**Review by nhat-nguyen — APPROVED** [review]

**Review by nhat-nguyen — APPROVED** [review]

**Comment by nhat-nguyen** [comment]
Merged! Thank @shintaro-iwasaki again!

**Pull request merged**
